File: scala/34.scala
import scala.annotation.tailrec
object Solution {
    def searchRange(nums: Array[Int], target: Int): Array[Int] = {
      @tailrec
      def first(i: Int, j: Int): Int = {
        if (i >= j) return if (nums(i) == target) i else -1
        val mid = i + (j-i)/2
        if (nums(mid) >= target) return first(i, mid)
        else return first(mid+1, j)
      }
      @tailrec
      def last(i: Int, j: Int): Int = {
        if (i >= j) return if (nums(i) == target) i else -1
        val mid = i + (j-i)/2+1
        if (nums(mid) <= target) return last(mid, j)
        else return last(i, mid-1)
      }
      if (nums.length == 0) return Array(-1, -1)
      val l = first(0, nums.length - 1)
      if (l == -1) return Array(-1,-1)
      Array(l, last(0, nums.length - 1))
    }
}

File: go/34.go
func searchRange(nums []int, target int) []int {
	first = -1
	last = -1
	findFirst(nums, 0, len(nums)-1, target)
	findLast(nums, 0, len(nums)-1, target)
	return []int{first, last}
}

var first = -1

func findFirst(nums []int, l, r, target int) {
	if l > r {
		return
	}
	mid := (r-l)/2 + l
	if nums[mid] >= target {
		if nums[mid] == target && (first == -1 || first > mid) {
			first = mid
		}
		findFirst(nums, l, mid-1, target)
	} else {
		findFirst(nums, mid+1, r, target)
	}
}

var last = -1

func findLast(nums []int, l, r, target int) {
	if l > r {
		return
	}
	mid := (r-l)/2 + l
	if nums[mid] <= target {
		if nums[mid] == target && (last == -1 || last < mid) {
			last = mid
		}
		findLast(nums, mid+1, r, target)
	} else {
		findLast(nums, l, mid-1, target)
	}
}